Send key events to flutter

maun opened the pull request:
Start of implementation for #72

Currently the code pretends to be android, also the scan code and modifiers are not sent. Ideally there would either be a matching platform in [flutter](https://github.com/flutter/flutter/blob/09276bea258737e11b109e227f70eac94a4e1691/packages/flutter/lib/src/services/raw_keyboard.dart#L134), or a generic/desktop platform.
It would also be a good idea to look at the iOS code base how they forward events to multiple plugins.

stuartmorgan commented on macos/library/FLEKeyEventPlugin.m:
```diff
@@ -0,0 +1,57 @@
+    NSDictionary* message = @{
+                                @"keymap": @"android",
+                                @"type": type,
+                                @"keyCode": [NSNumber numberWithUnsignedShort: event.keyCode],
```
That value can just be `@(event.keyCode)`

maun commented on macos/library/FLEViewController.h:
```diff
@@ -71,12 +76,20 @@
+ * Sends a platform message to the Flutter engine on the given channel.
+ * @param message The raw message.
+ */
+- (void)dispatchMessage:(nonnull NSDictionary*)message onChannel:(nonnull NSString *)channel;
```
There are flutter plugins defining and using [EventChannels](https://docs.flutter.io/javadoc/io/flutter/plugin/common/EventChannel.html) which as I understand are just raw messages. It looks like PlattformMessages [were changed to](https://github.com/flutter/flutter/wiki/Upgrading-Flutter-projects-from-using-PlatformMessages-to-using-channels) FlutterMessageChannel and FlutterMethodChannel. Maybe something similar should be done here.

maun commented on macos/library/FLEKeyEventPlugin.m:
```diff
@@ -0,0 +1,57 @@
+}
+
+- (void)dispatchKeyEvent:(NSString *)type event:(NSEvent *)event {
+    // TODO: Do not use 'android'
```
I added the issue link to the TODO comment.

Using "fuchsia" instead of android would change the class of the `data` property of the KeyEvent. If we change this handlers in the dart code would need to be adapted. Currently the handler in the example expects `RawKeyEventDataAndroid`. Also I am not filling all expected properties of the data class.
The [Fuchsia](https://github.com/flutter/flutter/blob/09276bea258737e11b109e227f70eac94a4e1691/packages/flutter/lib/src/services/raw_keyboard.dart#L83) implementation has no `keyCode` field, but instead `hidUsage`, not sure how to convert this.

stuartmorgan replied:
> Also I am not filling all expected properties of the data class.

Is there a reason you're not passing modifier key state via metaState? That seems like it would be necessary data for many uses.

stuartmorgan replied:
This wasn't resolved, at it still seems like passing modifier keys will be important, but I'm okay with leaving that for follow-up work.

maun commented on macos/library/FLEViewController.m:
```diff
@@ -383,25 +393,50 @@ - (void)initPlugins {
+#pragma mark - NSResponder key events
+
+- (void)keyDown:(NSEvent *)event {
+    for (NSResponder* responder in self.additionalKeyResponders) {
```
This seems to work as is, when using text input and key events the order is as follows:
* KeyDown
* Text
* KeyUp

Because the keyevent plugin is added first it handles the event first. I changed the NSMutableSet to NSOrderedMutableSet to guarantee this.

maun commented:
Thanks for your review @stuartmorgan. Can you recheck the PR? The biggest problem is that I misuse the android key event data, not even including all properties, but I do not know how to solve this. Also added some answers to your comments but GitHub hides them because i changed the code.

In the example there seems to be a minor issue: Selecting the text field, then changing to the new page and back, the text field looks like it has focus but you have to select it again to type something. Not sure why requesting the focus does not unfocus the old one.

stuartmorgan commented on example_flutter/lib/keyboard_test_page.dart:
```diff
@@ -0,0 +1,85 @@
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/services.dart';
+
+class KeyboardTestPage extends StatefulWidget {
```
Please document at least the public class (which is an analyzer failure under the current settings).

stuartmorgan commented on example_flutter/lib/keyboard_test_page.dart:
```diff
@@ -0,0 +1,85 @@
+  @override
+  Widget build(BuildContext context) {
+    return new Scaffold(
+      appBar: AppBar(
```
It would be nice to give this a title to provide some context.

stuartmorgan commented on example_flutter/lib/keyboard_test_page.dart:
```diff
@@ -0,0 +1,85 @@
+    );
+  }
+
+  void onKeyEvent(RawKeyEvent event) async {
```
Why is this async?

stuartmorgan commented on example_flutter/lib/keyboard_test_page.dart:
```diff
@@ -0,0 +1,85 @@
+
+  void onKeyEvent(RawKeyEvent event) async {
+    int keyCode;
+    bool isKeyDown;
```
Prefer to declare variables as close to their first use as possible, so keyCode attached to the second switch below, and this attached to the first switch.

stuartmorgan commented on example_flutter/lib/main.dart:
```diff
@@ -170,6 +171,10 @@ class _MyHomePage extends StatelessWidget {
             ),
             TextInputTestWidget(),
             FileChooserTestWidget(),
+            new RaisedButton(child: new Text('Test raw keyboard events'), onPressed: () {
```
Please dartfmt this file; this line is too long.

stuartmorgan commented on macos/library/FLEKeyEventPlugin.m:
```diff
@@ -0,0 +1,54 @@
+
+#import "FLEViewController.h"
+#import "FLEViewController+Internal.h"
+#import "FLEKeyEventPlugin.h"
```
This should be the first import, with a blank line after it, per style guide.

stuartmorgan commented on macos/library/FLEKeyEventPlugin.m:
```diff
@@ -0,0 +1,54 @@
+- (void)dispatchKeyEvent:(NSString *)type event:(NSEvent *)event {
+  // TODO: Do not use 'android'. Reconsider when https://github.com/google/flutter-desktop-embedding/issues/47 is solved.
+  NSDictionary* message = @{
+                            @"keymap": @"android",
```
The new code all still needs to be run through clang-format. If you don't have it installed let me know and I can do that as a final pass.

stuartmorgan commented on macos/library/FLEKeyEventPlugin.m:
```diff
@@ -0,0 +1,54 @@
+  result(FLEMethodNotImplemented);
+}
+
+- (void)dispatchKeyEvent:(NSString *)type event:(NSEvent *)event {
```
This needs a declaration comment (per style guide). It should also follow ObjC naming conventions that the word before the : names the param, so dispatchKeyEvent:(...)event ofType:(...)type

stuartmorgan commented on macos/library/FLEViewController+Internal.h:
```diff
@@ -0,0 +1,23 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
This should import FLViewController.h

stuartmorgan commented on macos/library/FLEViewController.h:
```diff
@@ -71,7 +71,19 @@
- * Sends a platform message to the Flutter engine on the given channel.
+ * Adds a responder for keyboard events. Key up and key down events are forwarded to all added responders.
+ */
+- (void)addKeyResponder:(nonnull NSResponder*)responder;
```
Per the first part of my comment about this change in the previous round, I don't think arbitrary third-party plugins should be able to be key event handlers, so these should be in the internal header.

stuartmorgan commented on macos/library/FLEViewController.m:
```diff
@@ -163,37 +170,49 @@ - (BOOL)addPlugin:(nonnull id<FLEPlugin>)plugin {
+    
+    FlutterResult result = FlutterEngineSendPlatformMessage(_engine, &platformMessage);
+    if (result != kSuccess) {
+        NSLog(@"Flutter engine send unsuccessful response for message %@: (%d).", message, result);
```
Typo: send->sent

Also, since the dictionary could be a lot to display, I'd put the error response before it:
Flutter engine sent unsuccessful response (%d) for message %@

stuartmorgan commented on macos/library/FLEKeyEventPlugin.m:
```diff
@@ -0,0 +1,54 @@
+                            @"keymap": @"android",
+                            @"type": type,
+                            @"keyCode": @(event.keyCode),
+                            };
```
Pulling this into a review comment since GitHub's UI makes it easy to miss replies to comments on outdated changes:

Is there a reason you're not passing modifier key state via metaState? That seems like it would be necessary data for many uses.

maun replied:
No particular reason. I did not need it and was not sure about the encoding.

stuartmorgan commented:
Is this still something you're interested in landing? The last review round is mostly just minor changes, so it's very close.

gedw99 commented:
@maun  Rearding https://github.com/google/flutter-desktop-embedding/pull/73#issuecomment-401825745,  you commented that you had trouble with key events. Did you ever get an answer about how to resolve this ?

maun commented:
@stuartmorgan I was rather busy before but now got around to address your review, sorry for the delay.

@gedw99 I did not look further into this and

stuartmorgan commented:
Thanks for the contribution, and sorry for the long delay!
